src/lib/device-os-version-util.js

```diff
@@ -0,0 +1,161 @@
+ * @param platformName - the platform name
+ * @param module - the module object to download
+ * @param baseUrl - the base url for the api
+ * @param version - the version to download
+ * @param progressBar - an instance of ProgressBar to use for progress updates
+ * @returns {Promise<string>} - the path to the binary
+ */
+const downloadBinary = async ({ platformName, module, baseUrl, version, progressBar }) => {
+	const binaryPath= getBinaryPath(version, platformName);
+	// fetch the binary
+	const uri = `${baseUrl}/${module.filename}`;
+	return downloadFile(uri, binaryPath, module.filename, progressBar);
+};
+
+const isModuleDownloaded = async (module, version, platformName) => {
+	// check if the module is already downloaded
+	const binaryPath = getBinaryPath(version, platformName);
+	const filePath = path.join(binaryPath, module.filename);
+	const exits = await fs.pathExists(filePath);
+	if (exits) {
+		const parser = new HalModuleParser();
+		const info = await parser.parseFile(filePath);
+		return info.crc.storedCrc === module.crc.storedCrc && info.crc.actualCrc === module.crc.actualCrc;
+	}
+	return false;
+};
+
+/**
+ * Download the binaries for the given platform and version by default the latest version is downloaded
+ * @param {Object} api - the api object
+ * @param {number} platformId - the platform id
+ * @param {string} version - the version to download (default: latest)
+ * @param {Object} ui - allow us to interact in the console
+ * @returns {Promise<*[]>} - true if successful
+ */
+const downloadDeviceOsVersionBinaries = async ({ api, platformId, version='latest', ui }) => {
+	try {
+		const downloadedBinaries = [];
+		// get platform by id from device-constants
+		const platform = Object.values(deviceConstants).filter(p => p.public).find(p => p.id === platformId);
+		// get the device os versions
+		const deviceOsVersion = await api.getDeviceOsVersions(platformId, version);
+		// download binaries for each module in the device os version
+		for await (const module of deviceOsVersion.modules) {
+			//TODO (hmontero) - make sure downloadedBinaries returns the full path to the binary
+			const isDownloaded = await isModuleDownloaded(module, deviceOsVersion.version, platform.name);
+			const binaryPath = getBinaryPath(deviceOsVersion.version, platform.name);
+			if (!isDownloaded) {
+				let progressBar;
+				// if is in silent mode don't create a progress bar
+				if (!ui.isOutputMuted()) {
+					progressBar = ui.createProgressBar(`Downloading ${module.filename}`);
+				} else {
+					ui.write(`Downloading ${module.filename}`);
+				}
+				const downloadedBinaryName = await downloadBinary({
+					platformName: platform.name,
+					module,
+					baseUrl: deviceOsVersion.base_url,
+					version: deviceOsVersion.version,
+					progressBar,
+				});
+				downloadedBinaries.push(path.join(binaryPath, downloadedBinaryName));
+			} else {
+				ui.write(`Skipping download of ${module.filename} as it is already downloaded`);
```

Comment:
No need for a message here as this will be the usual case after the first download.